### PR TITLE
Support for splitting with global variables.

### PR DESCRIPTION
**Pull request by alex4o**
Works when the global variable's initializer doesn't depend on other
GlobalVariables or Constants. To make this work the semantics of the
globals needs to be changed, which may introduce subtle bugs. Every
output bitcode file can be build to a .so file. Also functions named after
already exisint libraries can cause problems like `c()` creates `libc.so`. 
Probably I should add an underscore to the name.

### CODE REVIEW

**Comment by ltratt** [comment]
Can you explain *how* this works (roughly)? Also, AFAICS, none of the tests actually test this new feature?

**Comment by alex4o** [comment]
It first creates a temporary bitcode file where all of the global variables are global. Then extract the global variables with their initiallizers. Then remove the initiallizers make all the functions global and extract all the functions. This is the basic premise. There is not test for the new feature yet. But that shouldn't be too hard to add.

**Comment by ltratt** [comment]
https://github.com/capablevms/llvm-function-split/pull/8/commits/3373a9bf56178d9c36dde95ac9bbb3a3a1db2429 says "add a test" but doesn't seem to add a test? :p

> It first creates a temporary bitcode file where all of the global variables are global. Then extract the global variables with their initiallizers. Then remove the initiallizers make all the functions global and extract all the functions.

The first two sentences make sense but not the third. What does "make all the functions global" and "extract all the functions" mean?

**Comment by alex4o** [comment]
This is what I am speaking about: https://github.com/alex4o/llvm-function-split/blob/main/.buildbot.sh#L20-L24

About the second point. It means that I go through all of the functions make them [ExternalLinkage](https://llvm.org/doxygen/classllvm_1_1GlobalValue.html#aedfa75f0c85c4aa85b257f066fbea57ca6c93794d7b99cd433e96c53eadb15a6e) and [DefaultVisibility](https://llvm.org/doxygen/classllvm_1_1GlobalValue.html#a9141f967188383108a69cc1b8ed3c195a2491e41d821f1d8fd3958ce3df2fddb2). Then save the files and extrac all of the functions from it.

**Comment by ltratt** [comment]
Can you point me at the `.c` file(s) which test the new functionality?

**Comment by alex4o** [comment]
Oh there is no new modifications in the c files. They are just compiled differently. Which is the point. First I test If it compiles into a single static file fine when split. Then I test if it compiles when all of the things are in different shared libraries. As compiling statically is a prerequisite for compiling into modules.

**Comment by ltratt** [comment]
If I understand correctly, some inputs would have broken before this PR but work now? If so, can we add a didn't-work-before-this-PR test to this PR?

**Comment by alex4o** [comment]
The test I added would have failed before, for this input [llvm-extract-static.c](https://github.com/alex4o/llvm-function-split/blob/main/tests/test-llvm-extract-static.c). Because It couldn't deal with the global variable before.

**Comment by ltratt** [comment]
> The test I added would have failed before

Out of interest: failed in what way?

**Comment by alex4o** [comment]
With a linker error: `relocation R_X86_64_PC32 against symbol`

**Comment by ltratt** [comment]
OK, got it.

Can you squash and at the same time update the commit message with more details so that future-me-and-you and remember what this PR did? :)

**Comment by alex4o** [comment]
like that?

**Comment by ltratt** [comment]
The first part of the new commit message looks good.

>This is why the globals first need to be modified.
>So that they are truly global, and then modified again to be externally
>initialized as to when they are moved with their respective function
>users the initialization code will be in a different module.

This doesn't quite parse. I think the first sentence should partly run into the first, the the second should be split into two?

>To make this work the
>semantics of the globals needs to be changed, which may introduce subtle
>bugs. 

Is this saying that this PR introduces subtle bugs or fixes subtle bugs?

**Comment by alex4o** [comment]
is that better?

**Comment by ltratt** [comment]
I still don't understand "In the case the original code depended on the assumption that these variables won't change it might not work correctly." It's the "might" bit I'm confused by: are we saying this PR solves *all* problems or *some* problems? If the latter, what problems doesn't it solve?

**Comment by alex4o** [comment]
This is the main problem I couldn't think of a problem, but I feel like there might be some cases where that might be problematic.

**Comment by ltratt** [comment]
> This is the main problem I couldn't think of a problem, but I feel like there might be some cases where that might be problematic.

The technical term for this is "fence sitting" :) We need to work this out (as best we can) one way or the other. Either we feel that we've tried all possibilities (bearing in mind that without a formal proof we can never be *100% sure*) or we know that there are possibilities yet to try.

For example, have we tried running this on code bases larger than Lua yet?

**Comment by cristianurlea** [comment]
>  Also functions named after already exisint libraries can cause problems like c() creates libc.so. Probably I should add an  underscore to the name. 

I don't think an underscore will cut it: at very least think of a different prefix maybe `c()` could become `seclib_c`. 
You may also get internal collisions, so perhaps look at what a proper name mangling implementation does: https://clang.llvm.org/doxygen/Mangle_8cpp_source.html

**Comment by alex4o** [comment]
I am looking at running it on cpython but it still doesn't work. The name mangling will be useful I agree. I read a little about how these things are handled internally and now I think there might not be a problem with the global variables. I will add that information in the commit message and we can look for a further pr with better mangling?

**Comment by ltratt** [comment]
> I am looking at running it on cpython but it still doesn't work.

Aha, so we *know* that there are problems -- there's no "might" about it. It's OK for us not to know exactly what the problem is, but we should state some cases where it doesn't work. Please force push an update.

**Comment by alex4o** [comment]
I changed that sentence, I am not sure if that is enough?

**Comment by ltratt** [comment]
Just to be clear: you don't know why CPython (notice the capitalisation!) doesn't work? Also the final sentence "The end result is that every output bitcode file can be build to a .so file." doesn't fit anymore.

**Comment by alex4o** [comment]
I  think that sound fine now?

**Comment by ltratt** [comment]
Let's just delete "The end result is that golbal variables without dependencies can be moved to new modules and they can be compiled to a .so files." and then we're done.

**Comment by alex4o** [comment]
bors try

**Comment by ltratt** [comment]
I'm afraid bors is down at the moment :/ When it's back up I'll r+ this.

**Comment by alex4o** [comment]
When I reading this I expected `I'm afraid` to finish with `I can't do that dave`

**Comment by alex4o** [comment]
bors try

**Comment by alex4o** [comment]
bors try

**Comment by alex4o** [comment]
bors try

**Comment by alex4o** [comment]
bors try

**Comment by alex4o** [comment]
bors try

**Comment by alex4o** [comment]
This is ready for squashing

**Comment by ltratt** [comment]
Can you briefly explain what the problem was?

**Comment by alex4o** [comment]
So I renamed the `int c()` function to `int cc()`, but the move function was still trying to move `c`. This is the short version.

**Comment by ltratt** [comment]
> but the move function was still trying to move c

How was it doing that if there wasn't a function still called `c`?

**Comment by alex4o** [comment]
The problem is that I don't look up the functions by strings. There is a function that gets called for each llvm::Function that returns the operation that has to happen. And when it sees `cc` it says `skip`.

**Comment by ltratt** [comment]
Why is `cc` special cased?

File: main.cpp [hunk 1]
```diff
@@ -164,7 +167,7 @@ int main(int argc, char **argv) {
             return FunctionAction::MakePrivate;
           } else if (function.getName().equals("b")) {
             return FunctionAction::MakePublic;
-          } else if (function.getName().equals("c")) {
+          } else if (function.getName().equals("cc")) {
             return FunctionAction::MakePublic;
           } else {
             return FunctionAction::Skip;
```
Comment (alex4o) [inline main.cpp:173]: It is just that I had renamed the function in the c file and it had to be renamed here too.

Comment (ltratt) [reply main.cpp:173]: Yes, but why is `cc` special cased (or "why is `c` special case) or whatever? I don't understand why we have any function names that are special cased.

Comment (alex4o) [reply main.cpp:173]: Oh because when moved and compiled `c` was becoming `libc.so`.

**Comment by alex4o** [comment]
Ok, this pr is for the last splitter `split.cpp`, in this repo. There are 3 splitters `main.cpp` the old splitter that manually used the llvm api to split. the `find-static.cpp` one that find's static functions and tries to move around them. And the latest one that moves everything `split.cpp`. I changed one of the tests that was shared between the split.cpp and main.cpp splitters.

**Comment by ltratt** [comment]
OK, can we be clear about the relationship between `main.cpp` and `split.cpp`? Are they totally different splitters? If so, should we add another commit which gives them clearly different names or ... ?

**Comment by alex4o** [comment]
Yes they are unrelated, but the main.cpp one was the first, as I thought there wouldn't be any changes to the structure of the project.

**Comment by ltratt** [comment]
https://github.com/capablevms/llvm-function-split/pull/8/commits/91bf53dbe58e9f3eb4bf9299d615fd4a8831fd2e helps -- good :) Can we add to the README (or the `cpp` files or somewhere) a brief description of the difference between the three splitters?

**Comment by alex4o** [comment]
https://github.com/capablevms/llvm-function-split/pull/8/commits/6c9e1cf8932fcb99241e64f8de3beb742f4a08b7 I hope that works.

**Comment by ltratt** [comment]
That helps! Which of the splitter(s) special cases functions and shouldn't be used for general purposes? Let's document that too.

**Comment by alex4o** [comment]
Done

File: Readme.md [hunk 2]
```diff
@@ -33,6 +33,7 @@ There are 2 other utilities that reside in this repository. The [first one](manu
 how splitting can be done using the llvm API instead of using `llvm-extract`.
 The [other](find-and-split-static.cpp) utility focuses on finding functions which are not public and should
 be. It also tries to find ways to split the bitcode files while preserving 
-the linkage of the functions.
+the linkage of the functions. These splitters are not meant for general purpose
```
Comment (ltratt) [inline Readme.md:36]: Which splitters are "these"?

**Comment by alex4o** [comment]
Better?

**Comment by ltratt** [comment]
Works for me. Please can you reformat the new text in the README to be a consistent line width with what's already there. Once that's done we're probably ready for squashing.

**Comment by alex4o** [comment]
The text is re-formatted

**Comment by ltratt** [comment]
Please squash.

**Comment by alex4o** [comment]
Like that?

**Comment by alex4o** [comment]
@ltratt uh sry, can you look into this?

**Pull request merged**